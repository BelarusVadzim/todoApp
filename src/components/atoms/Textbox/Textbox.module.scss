@import "src/variables.scss";

.textbox {
  font-size: $font-size;
  color: $primary-font-color;
  font-family: $main-font;
  caret-color: $primary-color;
  background-color: transparent;
  outline: none;
  border: none;
  width: 100%;
  display: flex;
}
