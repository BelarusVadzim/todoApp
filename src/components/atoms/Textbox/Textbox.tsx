import React, { KeyboardEvent, ChangeEvent, useState } from 'react';
import style from './Textbox.module.scss';

type TextboxProps = {
  onReturn?: (text: string) => void;
};

const EnterKey = 'Enter';

const Textbox: React.FC<TextboxProps> = ({ onReturn }) => {
  const [text, setText] = useState('');

  const keyDownHandler = (
    keyboardEventArg: KeyboardEvent<HTMLInputElement>
  ) => {
    if (EnterKey === keyboardEventArg.key) {
      if (onReturn) onReturn(keyboardEventArg.currentTarget.value);
      setText('');
    }
  };

  const changeHandler = (changeEventArg: ChangeEvent<HTMLInputElement>) => {
    setText(changeEventArg.target.value);
  };

  return (
    <input
      type="text"
      className={style.textbox}
      onChange={changeHandler}
      onKeyDown={keyDownHandler}
      value={text}
      maxLength={50}
    />
  );
};

export default Textbox;
