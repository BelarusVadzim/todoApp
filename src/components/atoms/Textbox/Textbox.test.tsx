import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Textbox from './Textbox';

const textValue = 'textValue';
jest.mock('./Textbox.module.scss', () => ({ textbox: 'textboxStyle' }));
let mockSetText: jest.Mock;
jest.mock('react', () => {
  mockSetText = jest.fn();
  const actual = jest.requireActual('react');
  return {
    ...actual,
    useState: jest.fn().mockReturnValue([textValue, mockSetText]),
  };
});

afterEach(() => {
  jest.clearAllMocks();
});

describe('<Textbox />', () => {
  it('should render properly', () => {
    const component = render(<Textbox />);

    expect(component.asFragment()).toMatchSnapshot();
  });

  it('when text typed witn out Enter key should call expected functions with expected values', () => {
    const onReturnHandler = jest.fn();
    const typedText = 'typedText';
    render(<Textbox onReturn={onReturnHandler} />);

    const inputElement = screen.getByRole('textbox');

    userEvent.type(inputElement, typedText);

    expect(onReturnHandler).toBeCalledTimes(0);
    expect(mockSetText).toBeCalledTimes(typedText.length);
  });

  it('when Enter key entered should call expected functions with expected values', () => {
    const onReturnHandler = jest.fn();
    const typedText = 'typedText';

    render(<Textbox onReturn={onReturnHandler} />);

    const inputElement = screen.getByRole('textbox');

    userEvent.type(inputElement, `${typedText}{Enter}`);

    expect(onReturnHandler).toBeCalledTimes(1);
    expect(onReturnHandler).toBeCalledWith(typedText);
    expect(mockSetText).toBeCalledTimes(typedText.length + 1);
    expect(mockSetText).toBeCalledWith('');
  });
});
