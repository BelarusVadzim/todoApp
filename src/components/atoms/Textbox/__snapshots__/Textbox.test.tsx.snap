// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Textbox /> should render properly 1`] = `
<DocumentFragment>
  <input
    class="textboxStyle"
    maxlength="50"
    type="text"
    value=""
  />
</DocumentFragment>
`;
